microprofile/WFLY-14932_mp_reactive_messaging_kafka_api.adoc

```diff
@@ -0,0 +1,105 @@
+
+SmallRye Reactive Messaging has had this functionality for a long time in its Kafka connector module, but as this mixes implementation details with public API, with no clear separation of what is what, it was decided not to introduce it until a suitable API was extracted.
+
+SmallRye Reactive Messaging has been enhanced to extract a Kafka user API in https://github.com/smallrye/smallrye-reactive-messaging/pull/1272. Note that this is a starting point, and more constructs are likely to end up in the Kafka API. As this is an external project, we have influence over what we want to add (as in I will add), but we don't have influence to stop them from adding what they see fit to add.
+
+== Issue Metadata
+
+=== Issue
+
+* https://issues.redhat.com/browse/WFLY-14932[WFLY-14932]
+
+=== Related Issues
+
+* https://issues.redhat.com/browse/EAP7-1666[EAP7-1666]
+* https://issues.redhat.com/browse/WFLY-14798[WFLY-14798 - Upgrade to MicroProfile Reactive Messaging 2.0]
+* https://issues.redhat.com/browse/WFLY-14800[WFLY-14800 - Modernise MicroProfile Reactive Messaging Quickstart for RM 2.0]
+
+=== Dev Contacts
+
+* mailto:{email}[{author}]
+
+=== QE Contacts
+TBD
+
+=== Testing By
+// Put an x in the relevant field to indicate if testing will be done by Engineering or QE. 
+// Discuss with QE during the Kickoff state to decide this
+* [ ] Engineering
+
+* [ ] QE
+
+=== Affected Projects or Components
+
+* WildFly - will need adjusting to add the new Kafka API maven module to the module containing the Kafka Connector
+* WildFly BOMs - the new API artifact will be added to these
+* SmallRye Reactive Messaging - contains the new API Maven module
+* WildFly Quickstarts - The Reactive Messaging Quickstart should be enhanced to show the Kafka API
+
+=== Other Interested Projects
+
+=== Relevant Installation Types
+// Remove the x next to the relevant field if the feature in question is not relevant
+// to that kind of WildFly installation
+* [x] Traditional standalone server (unzipped or provisioned by Galleon)
+
+* [ ] Managed domain
+
+* [x] OpenShift s2i
+
+* [x] Bootable jar
+
+== Requirements
+
+=== Hard Requirements
+The Kafka user API contains the following classes, the below shows what they allow you to access
+
+* https://github.com/smallrye/smallrye-reactive-messaging/pull/1272/files#diff-78e354d9794932a1d2b42b4c85153068ae53109160fe69218366a681d608f498[KafkaMetadataUtil] - contains utility methods to set `OutgoingKafkaRecordMetadata` in a Reactive Messaging `Message` to configure how messages are sent, and to read `IncomingKafkaRecordMetadata` from a `Message` received from Kafka to get information relating to Kafka. This utility class is present to wrap usage of `Metadata` which has made its way into the spec interfaces in the SmallRye Reactive Messaging API Maven module, while not yet part of the Microprofile Reactive Messaging specification.
```

Comment:
Hope that since this is not part of an official MP RM spec, won't cause any issue to us in the future.

Reply:
SmallRye Reactive Messaging relies on this mechanism internally so as long as we don't change the implementation for something else (which is **very** unlikely) we should be good.